modin/pandas/indexing.py

```diff
@@ -175,6 +201,32 @@ def is_range_like(obj):
     )
 
 
+def boolean_mask_to_numeric(indexer):
+    """
+    Convert boolean mask to numeric indices.
+
+    Parameters
+    ----------
+    indexer : list-like of booleans
+
+    Returns
+    -------
+    np.ndarray of ints
+        Numerical positions of ``True`` elements in the passed `indexer`.
+    """
+    if isinstance(indexer, (np.ndarray, Series, pandas.Series)):
+        return np.where(indexer)[0]
+    else:
+        # It's faster to build the resulting numpy array from the reduced amount of data via
+        # `compress` iterator than convert non-numpy-like `indexer` to numpy and apply `np.where`.
```

Comment:
Although `np.fromiter` probably suffers from reallocations when building an array from an iterator with an unknown size, it's faster indeed:
```
np.where(python_list):  3.7251316179754212
np.fromiter(compress):  2.688922952977009
```

<details><summary>Reproducer</summary>

```python
import numpy as np
import timeit
from itertools import compress

N = 10_000_000
NRUNS = 5

python_list = [True] * N

print(
    "np.where(python_list): ",
    timeit.timeit(lambda: np.where(python_list), number=NRUNS),
)
print(
    "np.fromiter(compress): ",
    timeit.timeit(
        lambda: np.fromiter(
            compress(data=range(len(python_list)), selectors=python_list),
            dtype=np.int64,
        ),
        number=NRUNS,
    ),
)
```

</details>